Feature/issue 202 vectorize all

rayleigh opened the pull request:
#### Submisison Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:
Vectorization of unary functions and tests of these vectorizations

#### Intended Effect:
This will add vectorized unary functions and their tests.

#### How to Verify:
Run unit tests

#### Side Effects:
Updated the existing testing vectorizing framework to skip tests if the expected value and test value are both NaN. The example function foo_fun now returns NaN for 0.

Added the appropriate `apply_scalar_unary` function to `stan/math/prim/mat.hpp`, `stan/math/rev/mat.hpp`, and `stan/math/fwd/mat.hpp`. All tests in `stan/math/mix/scal/fun` and `test/unit/math/mix/scal/prob/normal_test.cpp` were updated to include `stan/math/mix/mat.hpp` so that they wouldn't error out.  

#### Documentation:
I have documented the code of the vectorized functions, but I don't think it's fully user-facing yet.

#### Reviewer Suggestions: 

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):
Rayleigh

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

bob-carpenter commented:
That submissions checklist is supposed to be the pre-submission checklist!

bob-carpenter commented:
And it needs the model tests.  The tests aren't even complete yet.  Maybe you want to remove this pull request until it's complete.

syclik commented:
The model tests will need to be in the Stan library.

(Bob, this is related to how we deal with a new feature across repos.)

On Fri, Apr 1, 2016 at 12:59 PM, Bob Carpenter <notifications@github.com>
wrote:

> And it needs the model tests. The tests aren't even complete yet. Maybe
> you want to remove this pull request until it's complete.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/stan-dev/math/pull/276#issuecomment-204470893>
>

rayleigh commented:
Okay. I'm closing the pull request and reopen it when it has everything.

rayleigh commented:
I ran all unit tests and make cpplint. There's one build/include error from the make cpplint report, but when I looked through an old thread on the Stan-dev discussion group, it sounded like Jenkins could help me better understand the report.

bob-carpenter commented:
Jenkins won't help you with cpplint (unless it shows up in
a different compiler).

What is the error?  If it's the standard libraries at the
end one, just move the built-in library includes to the end.

- Bob

> On Apr 3, 2016, at 10:15 PM, rayleigh <notifications@github.com> wrote:
> 
> I ran all unit tests and make cpplint. There's one build/include error from the make cpplint report, but when I looked through an old thread on the Stan-dev discussion group, it sounded like Jenkins could help me better understand the report.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>

rayleigh commented:
Jenkins actually helped. I accidentally included `stan/math/prim/mat/vectorize/apply_scalar_unary.hpp` twice in `log1p_exp.hpp` instead of including it and the non-vectorized version.

syclik commented:
It looks like the current problem is with a compiler error:
http://d1m1s1b1.stat.columbia.edu:8080/job/Math%20Pull%20Request%20-%20Tests%20-%20Header/257/warnings18Result/new/

You need to include cmath in order for that to be fixed.



On Sun, Apr 3, 2016 at 11:39 PM, rayleigh <notifications@github.com> wrote:

> Jenkins actually helped. I accidentally included
> stan/math/prim/mat/vectorize/apply_scalar_unary.hpp twice in log1p_exp.hpp
> instead of including it and the non-vectorized version.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/stan-dev/math/pull/276#issuecomment-205120037>
>

rayleigh commented:
Got it. Thanks for pointing that out; I fixed a few other files with a similar mistake.

bob-carpenter commented:
@rayleigh  If you could get the pieces that are working checked in, we can review and merge and split the remaining ones out into their own issues.

rayleigh commented:
Okay. This pull request is only composed of vectorized functions that work; I've pulled out all functions that had issues when vectorizing them from this pull request. It touches a lot more files because I had to update the apply_scalar_unary include in stan/math/prim/mat.hpp, stan/math/rev/mat.hpp, and stan/math/fwd/mat.hpp. As a result, I updated all tests in stan/math/mix/scal/fun and test/unit/math/mix/scal/prob/normal_test.cpp to include stan/math/mix/mat.hpp so that they wouldn't error out.

rayleigh commented:
I know that you're probably busy, but have you had a chance to review this pull request? It only has the vectorized functions that work.

bob-carpenter commented:
Nope.  Totally slammed with a gazillion things and I've
been sick for a week.  I'll try to get to it over the
weekend.  I don't foresee any major problems given what's
been going on so far.

If you want to keep working on the next bit, you can
branch from your branch for the next chunk of work.

- Bob

> On Apr 15, 2016, at 12:50 PM, rayleigh <notifications@github.com> wrote:
> 
> I know that you're probably busy, but have you had a chance to review this pull request? It only has the vectorized functions that work.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>

syclik commented on stan/math/prim/mat.hpp:
```diff
@@ -46,6 +46,8 @@
 #include <stan/math/prim/mat/err/constraint_tolerance.hpp>
 #include <stan/math/prim/mat/err/validate_non_negative_index.hpp>
 
+#include <stan/math/prim/mat/vectorize/apply_scalar_unary.hpp>
```
I believe the includes here are alphabetical after the meta traits. Is there a reason this needs to be included before the `stan/math/prim/mat/fun/` files? If not, could we move it to alphabetical order?

rayleigh replied:
It needs to be before the `stan/math/prim/mat/fun/` files to avoid the exp error that we got earlier. After all, the tests for the vectorized functions include `stan/math/prim/mat/fun/mat.hpp` first because of the order of headers.

syclik replied:
that's indicating some sort of problem and we should work through it. This isn't something we should ignore. Can you hunt down which of the files force this error? Finding one would be useful and we can work through that.

rayleigh replied:
Okay. I tried changing the order for the prim version of `apply_scalar_unary.hpp` and the tests ran fine for the vectorized tests of `exp` and `foo`. However, when I tried changing the order for the rev version of `apply_scalar_unary`, the vectorized tests of `foo` crashed.

bob-carpenter replied:
What do you mean by "crashed"?  Failed to compile or
failed at run time or what?

> On Apr 21, 2016, at 12:18 PM, rayleigh <notifications@github.com> wrote:
> 
> In stan/math/prim/mat.hpp:
> 
> > @@ -46,6 +46,8 @@
> >  #include <stan/math/prim/mat/err/constraint_tolerance.hpp>
> >  #include <stan/math/prim/mat/err/validate_non_negative_index.hpp>
> >  
> > +#include <stan/math/prim/mat/vectorize/apply_scalar_unary.hpp>
> 
> Okay. I tried changing the order for the prim version of apply_scalar_unary.hpp and the tests ran fine for the vectorized tests of exp and foo. However, when I tried changing the order for the rev version of apply_scalar_unary, the vectorized tests of foo crashed.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>

rayleigh replied:
It failed to compile when I moved `stan/math/rev/mat/vectorize/apply_scalar_unary.hpp` after `stan/math/prim/mat.hpp` in `stan/math/rev/mat.hpp` . I get the following error messages, which look like the messages that I got earlier when trying to vectorize exp.hpp:

```
In file included from test/unit/math/mix/mat/vectorize/foo_test.cpp:1:
In file included from ./stan/math/mix/mat.hpp:12:
In file included from ./stan/math/prim/mat.hpp:91:
In file included from ./stan/math/prim/mat/fun/exp.hpp:4:
./stan/math/prim/mat/vectorize/apply_scalar_unary.hpp:40:41: error: implicit
      instantiation of undefined template
      'Eigen::internal::traits<stan::math::var>'
      typedef typename Eigen::internal::traits<T>::Scalar scalar_t;
                                        ^
./stan/math/prim/mat/fun/log.hpp:31:21: note: in instantiation of template class
      'stan::math::apply_scalar_unary<stan::math::log_fun, stan::math::var>'
      requested here
    inline typename apply_scalar_unary<log_fun, T>::return_t
                    ^
./stan/math/rev/scal/fun/grad_inc_beta.hpp:26:16: note: while substituting
      deduced template arguments into function template 'log' [with T =
      stan::math::var]
      var c1 = log(z);
               ^
lib/eigen_3.2.8/Eigen/src/Core/util/ForwardDeclarations.h:17:29: note: template
      is declared here
template<typename T> struct traits;
                            ^
In file included from test/unit/math/mix/mat/vectorize/foo_test.cpp:1:
In file included from ./stan/math/mix/mat.hpp:12:
In file included from ./stan/math/prim/mat.hpp:91:
In file included from ./stan/math/prim/mat/fun/exp.hpp:4:
./stan/math/prim/mat/vectorize/apply_scalar_unary.hpp:46:42: error: no member
      named 'RowsAtCompileTime' in 'stan::math::var'
      typedef Eigen::Matrix<scalar_t, T::RowsAtCompileTime,
                                      ~~~^
./stan/math/prim/mat/vectorize/apply_scalar_unary.hpp:40:41: error: implicit
      instantiation of undefined template
      'Eigen::internal::traits<stan::math::var>'
      typedef typename Eigen::internal::traits<T>::Scalar scalar_t;
                                        ^
./stan/math/prim/mat/fun/exp.hpp:31:21: note: in instantiation of template class
      'stan::math::apply_scalar_unary<stan::math::exp_fun, stan::math::var>'
      requested here
    inline typename apply_scalar_unary<exp_fun, T>::return_t
                    ^
./stan/math/rev/scal/fun/grad_inc_beta.hpp:28:16: note: while substituting
      deduced template arguments into function template 'exp' [with T =
      stan::math::var]
      var c3 = exp(lbeta(a, b)) * inc_beta(a, b, z);
               ^
lib/eigen_3.2.8/Eigen/src/Core/util/ForwardDeclarations.h:17:29: note: template
      is declared here
template<typename T> struct traits;
                            ^
In file included from test/unit/math/mix/mat/vectorize/foo_test.cpp:1:
In file included from ./stan/math/mix/mat.hpp:12:
In file included from ./stan/math/prim/mat.hpp:91:
In file included from ./stan/math/prim/mat/fun/exp.hpp:4:
./stan/math/prim/mat/vectorize/apply_scalar_unary.hpp:46:42: error: no member
      named 'RowsAtCompileTime' in 'stan::math::var'
      typedef Eigen::Matrix<scalar_t, T::RowsAtCompileTime,
                                      ~~~^
4 errors generated.
make: *** [test/unit/math/mix/mat/vectorize/foo_test.o] Error 1
make test/unit/math/mix/mat/vectorize/foo_test failed
exit now (04/21/16 14:48:06 EDT)
```

bob-carpenter replied:
I think that means `stan/math/rev/mat/vectorize/apply_scalar_unary.hpp`
should be including in the Eigen traits before anything else.

> On Apr 21, 2016, at 2:50 PM, rayleigh <notifications@github.com> wrote:
> 
> In stan/math/prim/mat.hpp:
> 
> > @@ -46,6 +46,8 @@
> >  #include <stan/math/prim/mat/err/constraint_tolerance.hpp>
> >  #include <stan/math/prim/mat/err/validate_non_negative_index.hpp>
> >  
> > +#include <stan/math/prim/mat/vectorize/apply_scalar_unary.hpp>
> 
> It failed to compile when I moved stan/math/rev/mat/vectorize/apply_scalar_unary.hpp after stan/math/prim/mat.hpp in stan/math/rev/mat.hpp . I get the following error messages, which look like the messages that I got earlier when trying to vectorize exp.hpp:
> 
> In file included from test/unit/math/mix/mat/vectorize/foo_test.cpp:1:
> In file included from ./stan/math/mix/mat.hpp:12:
> In file included from ./stan/math/prim/mat.hpp:91:
> In file included from ./stan/math/prim/mat/fun/exp.hpp:4:
> ./stan/math/prim/mat/vectorize/apply_scalar_unary.hpp:40:41: error: implicit
>       instantiation of undefined template
>       'Eigen::internal::traits<stan::math::var>'
>       typedef typename Eigen::internal::traits<T>::Scalar scalar_t;
>                                         ^
> ./stan/math/prim/mat/fun/log.hpp:31:21: note: in instantiation of template class
>       'stan::math::apply_scalar_unary<stan::math::log_fun, stan::math::var>'
>       requested here
>     inline typename apply_scalar_unary<log_fun, T>::return_t
>                     ^
> ./stan/math/rev/scal/fun/grad_inc_beta.hpp:26:16: note: while substituting
>       deduced template arguments into function template 'log' [with T =
>       stan::math::var]
>       var c1 = log(z);
>                ^
> lib/eigen_3.2.8/Eigen/src/Core/util/ForwardDeclarations.h:17:29: note: template
>       is declared here
> template<typename T> struct traits;
>                             ^
> In file included from test/unit/math/mix/mat/vectorize/foo_test.cpp:1:
> In file included from ./stan/math/mix/mat.hpp:12:
> In file included from ./stan/math/prim/mat.hpp:91:
> In file included from ./stan/math/prim/mat/fun/exp.hpp:4:
> ./stan/math/prim/mat/vectorize/apply_scalar_unary.hpp:46:42: error: no member
>       named 'RowsAtCompileTime' in 'stan::math::var'
>       typedef Eigen::Matrix<scalar_t, T::RowsAtCompileTime,
>                                       ~~~^
> ./stan/math/prim/mat/vectorize/apply_scalar_unary.hpp:40:41: error: implicit
>       instantiation of undefined template
>       'Eigen::internal::traits<stan::math::var>'
>       typedef typename Eigen::internal::traits<T>::Scalar scalar_t;
>                                         ^
> ./stan/math/prim/mat/fun/exp.hpp:31:21: note: in instantiation of template class
>       'stan::math::apply_scalar_unary<stan::math::exp_fun, stan::math::var>'
>       requested here
>     inline typename apply_scalar_unary<exp_fun, T>::return_t
>                     ^
> ./stan/math/rev/scal/fun/grad_inc_beta.hpp:28:16: note: while substituting
>       deduced template arguments into function template 'exp' [with T =
>       stan::math::var]
>       var c3 = exp(lbeta(a, b)) * inc_beta(a, b, z);
>                ^
> lib/eigen_3.2.8/Eigen/src/Core/util/ForwardDeclarations.h:17:29: note: template
>       is declared here
> template<typename T> struct traits;
>                             ^
> In file included from test/unit/math/mix/mat/vectorize/foo_test.cpp:1:
> In file included from ./stan/math/mix/mat.hpp:12:
> In file included from ./stan/math/prim/mat.hpp:91:
> In file included from ./stan/math/prim/mat/fun/exp.hpp:4:
> ./stan/math/prim/mat/vectorize/apply_scalar_unary.hpp:46:42: error: no member
>       named 'RowsAtCompileTime' in 'stan::math::var'
>       typedef Eigen::Matrix<scalar_t, T::RowsAtCompileTime,
>                                       ~~~^
> 4 errors generated.
> make: *** [test/unit/math/mix/mat/vectorize/foo_test.o] Error 1
> make test/unit/math/mix/mat/vectorize/foo_test failed
> exit now (04/21/16 14:48:06 EDT)
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>

rayleigh replied:
Okay. I compared this discussion to the one on the issue tracker: https://github.com/stan-dev/math/issues/202. They look very similar and this link from the tracker ultimately helped me understand it: http://stackoverflow.com/questions/14626033/implicit-instantiation-of-undefined-template-class.

syclik commented on stan/math/prim/mat/fun/abs.hpp:
```diff
@@ -0,0 +1,39 @@
+    struct abs_fun {
+      template <typename T>
+      static inline T fun(const T& x) {
+        using std::fabs;
```
Is this supposed to be `std::abs`?

bob-carpenter replied:
That include isn't doing anything.  But I don't think there is a `std::abs`.  If this is trying to wrap `abs()` as the doc says, you want `using stan::math::abs`.

rayleigh replied:
Got it. I made the changes.

syclik commented on stan/math/prim/mat/fun/acos.hpp:
```diff
@@ -0,0 +1,39 @@
+
+    /**
+     * Vectorized version of acos().
+     * @param x Container of variables in range [-1, 1].
```
is this comment correct? what happens when x is outside that range?

syclik replied:
based on the test, it looks like it's fine if x is outside the range. Update the comment?

rayleigh replied:
I was copying over the description from the manual. Should we update the manual or the comment?

bob-carpenter replied:
Both should match what the code actually does.

> On Apr 20, 2016, at 3:10 PM, rayleigh <notifications@github.com> wrote:
> 
> In stan/math/prim/mat/fun/acos.hpp:
> 
> > +     * Structure to wrap acos() so it can be vectorized.
> > +     * @param x Variable in range [-1, 1].
> > +     * @tparam T Variable type.
> > +     * @return Arc cosine of variable in radians. 
> > +     */
> > +    struct acos_fun {
> > +      template <typename T>
> > +      static inline T fun(const T& x) {
> > +        using std::acos;
> > +        return acos(x);
> > +      }
> > +    };
> > +
> > +    /**
> > +     * Vectorized version of acos().
> > +     * @param x Container of variables in range [-1, 1].
> 
> I was copying over the description from the manual. Should we update the manual or the comment?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>

rayleigh replied:
Okay. I changed the comment in my code.

syclik commented on stan/math/prim/mat/fun/asin.hpp:
```diff
@@ -0,0 +1,39 @@
+
+    /**
+     * Vectorized version of asin().
+     * @param x Container of variables in range [-1, 1].
```
update comment regarding range

rayleigh replied:
Okay. I changed the comment in my code.

syclik commented on stan/math/prim/mat/fun/cbrt.hpp:
```diff
@@ -0,0 +1,40 @@
+
+#include <stan/math/prim/mat/vectorize/apply_scalar_unary.hpp>
+#include <boost/math/special_functions/cbrt.hpp>
+#include <cmath>
```
remove `#include <cmath>` since it's not needed for `cbrt`

syclik commented on stan/math/prim/mat/fun/inv_Phi.hpp:
```diff
@@ -0,0 +1,44 @@
+      static inline T fun(const T& x) {
+        using stan::math::inv_Phi;
+        stan::math::check_bounded<T, T, T>(
+          "inv_Phi vectorize", "Probability variable", x, 0, 1);
```
The types here are incorrect for `check_bounded`, but I think we should be able to simplify this:
```
stan::math::check_bounded("inv_Phi", "Probability variable", x, 0.0, 1.0);
```
The types will be inferred properly here and we shouldn't promote 0 and 1 to `T`.
Note: I'm also changing the function name from "inv_Phi vectorize" to "inv_Phi" to match what the user will see.

syclik replied:
wait... why is this check here? Why is it not inside the `inv_Phi` function itself?

rayleigh replied:
Removed the check as you suggested.

syclik commented on stan/math/prim/scal/fun/Phi.hpp:
```diff
@@ -26,9 +25,7 @@ namespace stan {
-    inline typename boost::math::tools::promote_args<T>::type
-    Phi(const T x) {
+    inline double Phi(const double x) {
       // overridden in fvar and var, so can hard-code boost versions
```
delete these comments.

rayleigh replied:
Done.

syclik commented on stan/math/rev/mat.hpp:
```diff
@@ -5,6 +5,7 @@
 #include <stan/math/rev/scal/meta/is_var.hpp>
 #include <stan/math/rev/scal/meta/partials_type.hpp>
 
+#include <stan/math/rev/mat/vectorize/apply_scalar_unary.hpp>
```
this include should go at the end, unless there's some wicked order dependency.... if there is, then we should figure out why.

rayleigh replied:
There is. I tried changing the order and the tests failed. I agree that it's weird because of how the mix mat.hpp includes the other mat.hpp files.

syclik commented on test/unit/math/fwd/mat/vectorize/expect_val_deriv_eq.hpp:
```diff
@@ -3,8 +3,11 @@
 
 static inline void expect_val_deriv_eq(double exp_var, double test_var) {
+  using stan::math::is_nan;
+  if (is_nan(exp_var) && is_nan(test_var)) return;
```
why do we return here? Shouldn't we flag this condition as an error?

Instead of `return;` I expect to see something like `FAIL() << "value and derivatives are NaN"`

rayleigh replied:
NaN doesn't fit really well within the testing framework that I've created:
1) It's not an error so any tests using EXPECT_THROW will fail.
2) A lot of cmath functions will return NaN instead of throwing errors.
3) There are functions like cbrt, which are defined for a value, but their derivatives aren't. 

So, the easiest way to deal with it is to check if both the expected values and test values are NaN. If they are, then it's at least consistent. If they aren't, then there's a problem.

syclik replied:
hi @rayleigh, we need to do something about NaNs. Was that a suggestion?

bob-carpenter replied:
I believe this is just a way to make sure if the correct
answer is NaN that the result is NaN.  For instance,
we need to make sure log(-1) returns NaN and as far as I know,
this is the only way to do it other than writing a more general
equality function.

- Bob


> On Apr 27, 2016, at 6:21 PM, Daniel Lee <notifications@github.com> wrote:
> 
> In test/unit/math/fwd/mat/vectorize/expect_val_deriv_eq.hpp:
> 
> >  
> >  static inline void expect_val_deriv_eq(double exp_var, double test_var) {
> > +  using stan::math::is_nan;
> > +  if (is_nan(exp_var) && is_nan(test_var)) return;
> 
> hi @rayleigh, we need to do something about NaNs. Was that a suggestion?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>

rayleigh replied:
Daniel, checking that both the expected and test values are NaN isn't a suggestion. It's what I ultimately implemented in this pull request. I did so by:
1. Using `return` to exit early from functions (the code above) if both expected and test values are NaN
2. Using `continue` in a loop if both expected and test values are NaN
3. If the two options above are not possible, push the check within an `if block`

bob-carpenter replied:
You could pull out something like a `same_value()` function so you
don't have to repeat the logic.

```
void expect_same(double x, double y) {
  if (isnan(x) && isnan(y)) return;
  EXPECT_EQ(x, y);
}
```

But I don't see how to do this test for expected NaN values any
other way.

> On Apr 28, 2016, at 3:32 PM, rayleigh <notifications@github.com> wrote:
> 
> In test/unit/math/fwd/mat/vectorize/expect_val_deriv_eq.hpp:
> 
> >  
> >  static inline void expect_val_deriv_eq(double exp_var, double test_var) {
> > +  using stan::math::is_nan;
> > +  if (is_nan(exp_var) && is_nan(test_var)) return;
> 
> Daniel, checking that both the expected and test values are NaN isn't a suggestion. It's what I ultimately implemented in this pull request. I did so by:
> 1. Using return to exit early from functions (the code above) if both expected and test values are NaN
> 2. Using continue in a loop if both expected and test values are NaN
> 3. If the two options above are not possible, push the check within an if block
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>

rayleigh replied:
Got it. I created an expect_val_eq function to do this and just pushed the changes.

syclik commented on test/unit/math/mix/mat/fun/Phi_test.cpp:
```diff
@@ -0,0 +1,97 @@
+  static R apply(const T& x) {
+    using stan::math::Phi;
+    return Phi(x);
+  }
```
I'm really confused now. Why do we need to write this apply function? How does this test framework test the vectorize framework?

It looks like this is testing the scalar version of the function. If that's the case, this test should be not be in the `mat` folder and should be in the `scal` folder.

syclik replied:
Never mind. Bob explained it to me.

syclik commented on test/unit/math/mix/scal/fun/F32_test.cpp:
```diff
@@ -1,4 +1,7 @@
 #include <stan/math/mix/scal.hpp>
+#include <stan/math/mix/mat.hpp>
```
1. there are 3 includes of `stan/math/mix/mat.hpp`
2. this test should not include `mat.hpp`. why does it need `mat.hpp`?

rayleigh replied:
The test also includes `test/unit/math/rev/mat/fun/util.hpp`, which includes `stan/math/rev/mat.hpp` and thus `stan/math/rev/mat/vectorize/apply_scalar_unary.hpp`. So, it needs the `fwd` version of `stan/math/rev/mat/vectorize/apply_scalar_unary.hpp` to not error out. I thought the consistent way to do so was to add `stan/math/mix/mat.hpp`. Should I add `stan/math/fwd/mat.hpp` instead?

syclik replied:
1. this file shouldn't include `test/unit/math/rev/mat/fun/util.hpp`. If it does, perhaps we should think about where this file goes and/or what it includes.
2. look at lines 2-5. There are 3 includes that are the same.

rayleigh replied:
Okay. Looking through the code, I think we could include `test/unit/math/rev/mat/fun/util.hpp` instead because the test only looks to be using AVEC.

rayleigh replied:
I made the changes and the `mix/scal/fun` and `mix/prob/fun` tests are still passing.

syclik commented on test/unit/math/mix/scal/fun/Phi_test.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <stan/math/mix/scal.hpp>
+#include <stan/math/mix/mat.hpp>
```
shouldn't include mat.hpp

syclik commented on test/unit/math/prim/mat/vectorize/expect_prim_matrix_value.hpp:
```diff
@@ -19,11 +21,10 @@ void expect_prim_matrix_value() {
   EXPECT_EQ(a.size(), fa.size());
-  for (int i = 0; i < fa.size(); ++i)
+  for (int i = 0; i < fa.size(); ++i) {
+    if (is_nan(F::apply_base(a(i))) && is_nan(fa(i))) continue;
```
why do we ignore nan? isn't this an error?

syclik replied:
@rayleigh, we really should do something about NaN here.

syclik commented on test/unit/math/prim/mat/vectorize/expect_prim_matrix_value.hpp:
```diff
@@ -34,9 +35,15 @@ void expect_prim_matrix_value() {
     EXPECT_EQ(d[i].size(), fd[i].size());
-    for (int j = 0; j < fd[i].size(); ++j)
+    for (int j = 0; j < fd[i].size(); ++j) {
+      if (is_nan(F::apply_base(d[i](j))) && is_nan(fd[i](j))) return;
```
more ignoring nan. I think this should be an error.

syclik commented on test/unit/math/prim/mat/vectorize/expect_prim_row_vector_value.hpp:
```diff
@@ -28,8 +32,10 @@ void expect_prim_row_vector_value() {
     EXPECT_EQ(d[i].size(), fd[i].size());
-    for (int j = 0; j < fd[i].size(); ++j)
+    for (int j = 0; j < fd[i].size(); ++j) {
+      if (is_nan(F::apply_base(d[i](j))) && is_nan(fd[i](j))) continue;
```
nan. same comment.

syclik commented on test/unit/math/prim/mat/vectorize/expect_prim_scalar_value.hpp:
```diff
@@ -3,18 +3,25 @@
+    double exp_v = F::apply_base(double(input));
     double v = F::template apply<double>(input);
-    EXPECT_FLOAT_EQ(F::apply_base(double(input)), v);
+    if (is_nan(exp_v) && is_nan(v)) continue;
```
nan. same comment.

syclik commented on test/unit/math/prim/mat/vectorize/expect_prim_std_vector_value.hpp:
```diff
@@ -3,16 +3,19 @@
   vector<double> fy = F::template apply<vector<double> >(valid_inputs);
   EXPECT_EQ(valid_inputs.size(), fy.size());
   for (size_t i = 0; i < valid_inputs.size(); ++i) {
+    if (is_nan(F::apply_base(valid_inputs[i])) && is_nan(fy[i])) continue; 
```
more nan

syclik commented on test/unit/math/prim/mat/vectorize/expect_prim_std_vector_value.hpp:
```diff
@@ -24,14 +27,18 @@ void expect_prim_std_vector_value() {
     EXPECT_EQ(z[i].size(), fz[i].size());
-    for (size_t j = 0; j < z[i].size(); ++j)
+    for (size_t j = 0; j < z[i].size(); ++j) {
+      if (is_nan(F::apply_base(z[i][j])) && is_nan(fz[i][j])) continue; 
```
nan

syclik commented on test/unit/math/prim/mat/vectorize/expect_prim_std_vector_value.hpp:
```diff
@@ -24,14 +27,18 @@ void expect_prim_std_vector_value() {
   vector<double> u2 = F::template apply<vector<double> >(int_valid_inputs);
   EXPECT_EQ(int_valid_inputs.size(), u2.size());
   for (size_t i = 0; i < int_valid_inputs.size(); ++i) {
+    if (is_nan(F::apply_base(int_valid_inputs[i])) && is_nan(u2[i]))
```
nan

syclik commented on test/unit/math/prim/mat/vectorize/expect_prim_vector_value.hpp:
```diff
@@ -4,19 +4,23 @@
   EXPECT_EQ(b.size(), fb.size());
-  for (int i = 0; i < fb.size(); ++i)
+  for (int i = 0; i < fb.size(); ++i) {
+    if (is_nan(F::apply_base(b(i))) && is_nan(fb(i))) continue;
```
nan

syclik commented on test/unit/math/prim/mat/vectorize/foo_fun.hpp:
```diff
@@ -10,13 +10,14 @@ namespace stan {
     /**
      * Example of how to define a functor for a vectorized function.
-     * The example includes a constrained version of exp().
+     * The example includes a constrained, punctured version of exp().
```
do we need `foo_fun` any more? This pull request should have plenty of examples now, I believe.

rayleigh replied:
I think so because I think it can be useful as a sandbox function for any changes to the unary vectorization framework or testing framework.

bob-carpenter commented:
Now that everything's passing, is this ready to go?

bob-carpenter commented on stan/math/prim/mat.hpp:
```diff
@@ -229,6 +229,8 @@
 #include <stan/math/prim/mat/prob/wishart_log.hpp>
 #include <stan/math/prim/mat/prob/wishart_rng.hpp>
 
+#include <stan/math/prim/mat/vectorize/apply_scalar_unary.hpp>
```
(fix later) We don't need blank lines between includes.  And where possible, I'd prefer to just alphabetize (using lexicographic ordering on the path) everything.

bob-carpenter commented on stan/math/prim/mat/fun/Phi.hpp:
```diff
@@ -0,0 +1,39 @@
+
+    /**
+     * Vectorized version of Phi().
+     * @param x Container.
```
(fix later) `@tparam` first, with blank line between last line of text and first @ thing.

syclik commented on test/unit/math/mix/scal/fun/Phi_approx_test.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <stan/math/mix/scal.hpp>
+#include <stan/math/mix/mat.hpp>
```
this shouldn't include a mix header.

bob-carpenter replied:
Thanks for going through all the includes!

- Bob

> On Apr 27, 2016, at 6:30 PM, Daniel Lee <notifications@github.com> wrote:
> 
> In test/unit/math/mix/scal/fun/Phi_approx_test.cpp:
> 
> > @@ -1,4 +1,5 @@
> >  #include <stan/math/mix/scal.hpp>
> > +#include <stan/math/mix/mat.hpp>
> 
> this shouldn't include a mix header.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
>

syclik commented on test/unit/math/mix/scal/fun/abs_test.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <stan/math/mix/scal.hpp>
+#include <stan/math/mix/mat.hpp>
```
this test shouldn't include a mix header.

syclik commented on test/unit/math/mix/scal/fun/acos_test.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <stan/math/mix/scal.hpp>
+#include <stan/math/mix/mat.hpp>
```
this test shouldn't include a mix header.

syclik commented on test/unit/math/mix/scal/fun/acosh_test.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <stan/math/mix/scal.hpp>
+#include <stan/math/mix/mat.hpp>
```
this test shouldn't include a mix header.

syclik commented on test/unit/math/mix/scal/fun/asin_test.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <stan/math/mix/scal.hpp>
+#include <stan/math/mix/mat.hpp>
```
this test shouldn't include a mix header.

syclik commented on test/unit/math/mix/scal/fun/asinh_test.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <stan/math/mix/scal.hpp>
+#include <stan/math/mix/mat.hpp>
```
this test shouldn't include a mix header.

syclik commented on test/unit/math/mix/scal/fun/atan2_test.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <stan/math/mix/scal.hpp>
+#include <stan/math/mix/mat.hpp>
```
this test shouldn't include a mix header.

syclik commented on test/unit/math/mix/scal/fun/atan_test.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <stan/math/mix/scal.hpp>
+#include <stan/math/mix/mat.hpp>
```
this test shouldn't include a mix header.

syclik commented on test/unit/math/mix/scal/fun/atanh_test.cpp:
```diff
@@ -1,4 +1,5 @@
 #include <stan/math/mix/scal.hpp>
+#include <stan/math/mix/mat.hpp>
```
@rayleigh... there are more of these includes.

bob-carpenter commented:
Is this ready to merge or if not, what's left to test?

rayleigh commented:
I made the changes that you suggested. I want to note that I moved the function `cgrad` from the rev arr version of util.hpp to the rev scal version because `test/unit/math/mix/scal/fun/log_mix_test.cpp` uses cgrad.

rtrangucci commented on test/unit/math/mix/mat/fun/acos_test.cpp:
```diff
@@ -0,0 +1,97 @@
+
+/**
+ * This is the structure for testing mock function acos (defined in the
+ * testing framework).  See README.txt for more instructions.
```
@rayleigh we need to remove README.txt, along with all the references to the README

rayleigh replied:
Okay. I got a chance to work on it today and made those changes.

bob-carpenter replied:
Does anyone still need me to code review this again?
I'm traveling and pretty booked this week.

- Bob

> On Jun 27, 2016, at 8:24 AM, rayleigh <notifications@github.com> wrote:
> 
> In test/unit/math/mix/mat/fun/acos_test.cpp:
> 
> > @@ -0,0 +1,97 @@
> > +#include <stan/math/mix/mat.hpp>
> > +#include <gtest/gtest.h>
> > +#include <test/unit/math/prim/mat/vectorize/prim_scalar_unary_test.hpp>
> > +#include <test/unit/math/rev/mat/vectorize/rev_scalar_unary_test.hpp>
> > +#include <test/unit/math/fwd/mat/vectorize/fwd_scalar_unary_test.hpp>
> > +#include <test/unit/math/mix/mat/vectorize/mix_scalar_unary_test.hpp>
> > +#include <stan/math/prim/mat/fun/acos.hpp>
> > +#include <test/unit/math/prim/mat/vectorize/vector_builder.hpp>
> > +
> > +/**
> > + * This is the structure for testing mock function acos (defined in the
> > + * testing framework).  See README.txt for more instructions.
> 
> Okay. I got a chance to work on it today and made those changes.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

rayleigh commented:
I know that I haven't been pushing this, but is there anything else that I need to do? I know that there are a few future issues, but I've changed the text, updated the rev arr util.hpp, and marked tests where both expected and test values are NaN as success.

syclik commented:
One thing I haven't seen yet is how this performs for the univariate functions pre and post. Unless I've missed it. If so, can you repost? I'm actually concerned about that. 

> On Jul 5, 2016, at 6:51 PM, rayleigh <notifications@github.com> wrote:
> 
> I know that I haven't been pushing this, but is there anything else that I need to do? I know that there are a few future issues, but I've changed the text, updated the rev arr util.hpp, and marked tests where both expected and test values are NaN as success.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

rayleigh commented:
Daniel, what part of the univariate function performance are you concerned about?

syclik commented:
Wall time. I just want to make sure we're not going to take a significant performance hit for the existing univariate functions. 

> On Jul 5, 2016, at 7:31 PM, rayleigh <notifications@github.com> wrote:
> 
> Daniel, what part of the univariate function performance are you concerned about?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

rayleigh commented:
I see. I didn't measure that because I wasn't aware that I needed to, but looking at `foo_test`, it takes 447 ms to run the whole tests. 437 ms is from running the mix `expect_values test`.

However, the vectorize code is essentially putting the univariate function within a for loop. There is no optimization within the vectorize code itself because I believe this project is to make the univariate functions easier to use when writing Stan code.

bob-carpenter commented:
Everything should get compiled away through the templates
and it's the autodiff time that's the killer, not the
function computes, but it can't hurt to test.

Do I need to review more code at this point?

- Bob

> On Jul 5, 2016, at 4:54 PM, Daniel Lee <notifications@github.com> wrote:
> 
> Wall time. I just want to make sure we're not going to take a significant performance hit for the existing univariate functions. 
> 
> > On Jul 5, 2016, at 7:31 PM, rayleigh <notifications@github.com> wrote:
> > 
> > Daniel, what part of the univariate function performance are you concerned about?
> > 
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> > 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

rayleigh commented:
If `test/unit/math/rev/scal/fun/util.hpp`, `test/unit/math/rev/arr/fun/util.hpp`, and `test/unit/math/prim/mat/vectorize/expect_val_eq.hpp` have been reviewed, then there really haven't been any other code changes since I've pushed all the easy to vectorize univariate functions up. 

If we want to test performance, what scenarios should I test these functions in? For instance, if my function is foo, would I need to run the following tests before and after vectorization:
```
foo(int x)
foo(double x)
foo(rev x)
foo(fwd x)
foo(mix x)
```

bob-carpenter commented:
If you want to be thorough, look at log() applied to

T in {var, double} for

  * T                                    (scalar)
  * std::vector<T>                       (array)
  * Eigen::Matrix<T, Dynamic, 1>         (vector)
  * Eigen::Matrix<T, Dynamic, Dynamic>   (matrix)

We didn't have any higher-order vectorization than that and
have never run speed tests on forward-mode.

It doesn't need to be elaborate or checked in.

Also, did all the non-C++ files get removed?

- Bob


> On Jul 5, 2016, at 5:35 PM, rayleigh <notifications@github.com> wrote:
> 
> If test/unit/math/rev/scal/fun/util.hpp, test/unit/math/rev/arr/fun/util.hpp, and test/unit/math/prim/mat/vectorize/expect_val_eq.hpp have been reviewed, then there really haven't been any other code changes since I've pushed all the easy to vectorize univariate functions up.
> 
> If we want to test performance, what scenarios should I test these functions in? For instance, if my function is foo, would I need to run the following tests before and after vectorization:
> 
> foo(int x)
> foo(double x)
> foo(rev x)
> foo(fwd x)
> foo(mix x)
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

rayleigh commented:
I believe that all the non-C++ files got removed.

As for speed tests, I didn't suggest any vectorizations because for many of the univariate functions, there wasn't a vectorized version. So, what would I use to get the pre-vectorization results?

syclik commented:
I just want to make sure the univariate functions for double and var are
the about the same speed as without these changes. I know that everything
*should* be compiled away, but I'll believe it when I see it.

I'd start simple:
- time one function a whole bunch of times before and after
- time a couple of Stan programs that use one of the functions before and
after. Set seeds to known values and make sure everything is identical.

If those show no increase in speed, I'm happy to believe that the compilers
actually did what they're supposed to. If there's more than a 2% time
increase, I'd need to sit and think about it.


On Tue, Jul 5, 2016 at 9:10 PM, rayleigh <notifications@github.com> wrote:

> I believe that all the non-C++ files got removed.
>
> As for speed tests, I didn't suggest any vectorizations because for many
> of the univariate functions, there wasn't a vectorized version. So, what
> would I use to get the pre-vectorization results?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/276#issuecomment-230646713>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAZ_F14puqos3_4lVMABt7KXiFayUJpRks5qSwCigaJpZM4H92K0>
> .
>

bob-carpenter commented:
> On Jul 5, 2016, at 5:08 PM, rayleigh <notifications@github.com> wrote:
> 
> I see. I didn't measure that because I wasn't aware that I needed to, but looking at foo_test, it takes 447 ms to run the whole tests. 437 ms is from running the mix expect_values test.
> 
> However, the vectorize code is essentially putting the univariate function within a for loop. There is no optimization within the vectorize code itself because I believe this project is to make the univariate functions easier to use when writing Stan code.

Daniel just wants to make sure that things aren't
slower now than they used to be.  They weren't previously
optimized, either.

So just run a Stan program in the develop branch and
compare to your branch (CmdStan is easiest for that if
you check out the Stan branches as submodules) for each
of the previously vectorized functions (not many).

- Bob

P.S.  I'm off to Ann Arbor next Thursday!  Side trip from
      visiting my parents.

bob-carpenter commented:
Yes, I was suggesting just doing the log() function.
I hope I sent it!   const references to following
argument types at some scale (say 100 or 1000
elements in the list).

  T
  vector<T>
  Matrix<T, -1, 1>
  Matrix<T, 1, -1>
  matrix<T, -1, -1>

for T in {double, var}

- Bob

> On Jul 5, 2016, at 9:49 PM, Daniel Lee <notifications@github.com> wrote:
> 
> I just want to make sure the univariate functions for double and var are
> the about the same speed as without these changes. I know that everything
> *should* be compiled away, but I'll believe it when I see it.
> 
> I'd start simple:
> - time one function a whole bunch of times before and after
> - time a couple of Stan programs that use one of the functions before and
> after. Set seeds to known values and make sure everything is identical.
> 
> If those show no increase in speed, I'm happy to believe that the compilers
> actually did what they're supposed to. If there's more than a 2% time
> increase, I'd need to sit and think about it.
> 
> 
> On Tue, Jul 5, 2016 at 9:10 PM, rayleigh <notifications@github.com> wrote:
> 
> > I believe that all the non-C++ files got removed.
> >
> > As for speed tests, I didn't suggest any vectorizations because for many
> > of the univariate functions, there wasn't a vectorized version. So, what
> > would I use to get the pre-vectorization results?
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/stan-dev/math/pull/276#issuecomment-230646713>, or mute
> > the thread
> > <https://github.com/notifications/unsubscribe/AAZ_F14puqos3_4lVMABt7KXiFayUJpRks5qSwCigaJpZM4H92K0>
> > .
> >
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

syclik commented:
Jenkins, retest this please.
